### ISSUE TITLE
`Fut.to_promise` doesn't handle rejected futures correctly

### ISSUE THREAD

**Issue by v-gb** [dbuenzli/brr#58]
Given a `Fut.t` whose underlying promise is rejected, `Fut.to_promise` creates a promise that never resolves, which then prevents whatever surrounding code from cleaning up and displaying said error.

Example :

```ocaml
let test str to_promise =
  ignore (
      Jv.Promise.then'
        (to_promise
           ~ok:Fun.id
           (Fut.bind
              (Fut.return "qwe")
              failwith))
        (fun x -> Brr.Console.log [str; "ok"; x]; Jv.Promise.resolve Jv.undefined)
        (fun x -> Brr.Console.log [str; "error"; x]; Jv.Promise.resolve Jv.undefined))

let () = test "Fut.to_promise" Fut.to_promise
(* Bad. prints neither "ok", nor "error".  Instead you get a console warning saying "uncaught in promise: ..." *)

let fut_to_promise
    : ok:('a -> Jv.t) -> 'a Fut.or_error -> Jv.Promise.t
  = fun ~ok f ->
  Jv.Promise.bind (Jv.get (Obj.magic f) "fut")
    (function
     | Ok v -> Jv.Promise.resolve (ok v)
     | Error e -> Jv.Promise.reject e)

let () = test "custom Fut.to_promise" fut_to_promise
(* Good. prints in the "error" branch *)
```

The implementation of the module says it's an invariant that the promise of a future is never rejected, but the invariant isn't maintained in the `bind` function (and `map`). Not sure if that's an implementation or documentation bug. I think it'd be more useful to change the implementation as above, rather than changing the documentation.

**Comment by dbuenzli** [comment]
I don't fully grasp the problem and what you want. 

Is the problem that `Fut.t` doesn't handle exceptions ?

**Comment by v-gb** [comment]
> I don't fully grasp the problem and what you want.

Ok, let me try again.

> Is the problem that Fut.t doesn't handle exceptions ?

It depends what "handle" means.

I simply want `Fut` to let exceptions pass through : if the parameter of `bind` raises, I expect to be able to get that exception later. For instance, use that exception this way:

```ocaml
let f () =
  let open Fut.Result_syntax in
  let* file_content = Brr.File.text ... in
  if whatever then failwith "...";
  Fut.ok ()

let f_promise () = (* create a binding for javascript *)
  Fut.to_promise (f ()) ~ok:whatever
```
and use it in javascript:
```js
try {
  await f_promise()
} catch(e) {
  do stuff here
}
```

My problem is that the `do stuff here` bit never runs, because the `await` never returns nor throws.

And that's because the promise never resolves, because `Fut.to_promise` is implemented as (in javascript, and simplifying the `{fut: ...}` wrapper and the type conversion which are not relevant here):

```js
function to_promise(fut) {
  return new Promise((resolve_ok, resolve_err) => { fut.then(function_that_matches_on_ocaml_result) })
}
```

If `fut` resolves to a rejection, the `fut.then` bit does nothing (no callback), so the exception propagates to the next promise. But since the chain of promises is over, the error gets logged, and the returned promise stays unresolved.

My suggested implementation of `Fut.to_promise` above boils down to this instead:

```js
function to_promise(fut) {
  return fut.then(function_that_matches_on_ocaml_result)
}
```

If `fut` resolves to a rejection, the `fut.then` bit does nothing (no callback), so the exception propagates. But this time, the chain of promise is not over: the exception propagates into the returned promise.

**Comment by dbuenzli** [comment]
Thank you for your clarification.

My feeling is that you are a bit relying on something that should not be done. 

That is in contrast to say `lwt`, `'a Fut.t` values are purposedly not made to work with exceptions. Any exceptional behaviour should be made to explicitly appear in the `'a`. 

So I rather see the Fut values you want to treat as the result of programming errors.

**Comment by v-gb** [comment]
In that case, it'd be a documentation bug, as the .mli says nothing about not raising (it says the underlying promises don't reject, but that could just as well mean that the implementation catches all exceptions).

**Comment by dbuenzli** [comment]
The documentation says:

> A future ['a Fut.t] is an undetermined value of type ['a] that
becomes determined at an arbitrary point in the future.

To me it's obvious that raising an exception is not determining a value. But I'll try to add something.

> but that could just as well mean that the implementation catches all exceptions)

You'd need a value for the exceptional cases.
